### PR TITLE
TASK: Refactor open graph and twitter card to afx

### PR DESCRIPTION
**Pull request by Sebobo**

### CODE REVIEW

File: Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion [hunk 1]
```diff
@@ -1,89 +1,36 @@
...
+    image = ${q(node).property('openGraphImage')}
+    siteName = ${q(site).property('titleOverride') || site.label}
+    locale = Neos.Seo:LangAttribute
+    url = Neos.Neos:NodeUri {
+        node = ${node}
+        absolute = true
+    }
+    image = Neos.Fusion:Case {
+        default {
+            @position = 'start'
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion]: Shouldn't this be `@position = 'end'`?

File: Resources/Private/Fusion/Metadata/TwitterCard.fusion [hunk 2]
```diff
@@ -1,78 +1,33 @@
...
-        tagName = 'meta'
-        attributes {
-            name = 'twitter:description'
-            content = ${q(node).property('twitterCardDescription') ? q(node).property('twitterCardDescription') : q(node).property('metaDescription')}
-        }
-        @if.isNotBlank = ${!String.isBlank((q(node).property('twitterCardDescription') ? q(node).property('twitterCardDescription') : q(node).property('metaDescription')))}
+prototype(Neos.Seo:TwitterCard) < prototype(Neos.Fusion:Component) {
+    type = ${q(node).property('twitterCardType') || 'summary'}
+    title = ${q(node).property('twitterCardTitle') || q(node).property('titleOverride') || q(node).property('title')}
+    description = ${q(node).property('twitterCardDescription') ? q(node).property('twitterCardDescription') : q(node).property('metaDescription')}
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/TwitterCard.fusion]: `${q(node).property('twitterCardDescription') || q(node).property('metaDescription')}`

File: Resources/Private/Fusion/Metadata/TwitterCard.fusion [hunk 3]
```diff
@@ -1,78 +1,33 @@
...
+        asset = Neos.Fusion:Case {
+            default {
+                @position = 'end'
+                condition = ${true}
+                renderer = ${q(node).property('twitterCardImage')}
             }
         }
+        maximumWidth = 2000
+        maximumHeight = 2000
         @if.isImageSet = ${q(node).property('twitterCardImage') != null}
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/TwitterCard.fusion]: `@if.isImageSet = ${q(node).property('twitterCardImage')}`

**Comment by Sebobo** [comment]
I'll wait a bit with continuing this as #107 should be merged first and then I'll rebase.

File: Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion (hunk 1 above)
Comment (jonnitto) [reply Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion]: Ah, I got it. For the fallback…

**Comment by Sebobo** [comment]
@jonnitto can be tested now and I included your feedback 👍

File: Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion [hunk 4]
```diff
@@ -1,90 +1,36 @@
...
+            condition = ${Type.instance(q(node).property('openGraphImage'), 'Neos\Media\Domain\Model\ImageInterface')}
+            renderer = ${q(node).property('openGraphImage')}
+        }
+    }
+
+    renderer = afx`
+        <meta @key="type" property="og:type" content={props.type}/>
+        <meta @key="title" property="og:title" content={props.title}/>
+        <meta @key="siteName" property="og:site_name" content={props.title}/>
+        <meta @key="locale" property="og:locale" content={props.locale} @if.hasLocale={props.locale}/>
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion]: `@if.has`

Comment (jonnitto) [reply Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion]: I think we don't need the repetition here…

File: Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion [hunk 5]
```diff
@@ -1,90 +1,36 @@
...
+        <meta @key="description" property="og:description" content={props.description} @if.hasDescription={props.description}/>
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion]: `@if.has`

File: Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion [hunk 6]
```diff
@@ -1,90 +1,36 @@
...
+        <meta @key="url" property="og:url" content={props.url}/>
+        <Neos.Fusion:Join @key="image" @if.hasImage={props.image}>
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion]: `@if.has`

File: Resources/Private/Fusion/Metadata/TwitterCard.fusion [hunk 7]
```diff
@@ -1,87 +1,42 @@
...
+                renderer = 'Neos.Seo:TwitterCard.SummaryCard'
             }
         }
+        @if.isImageSet = ${this.asset}
     }
+
+    renderer = afx`
+        <meta @key="type" name="twitter:card" content={props.type}/>
+        <meta @key="title" name="twitter:title" content={props.title}/>
+        <meta @key="siteHandle" name="twitter:site" content={props.siteHandle} @if.hasSiteHandle={props.siteHandle}/>
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/TwitterCard.fusion]: `@if.has`

File: Resources/Private/Fusion/Metadata/TwitterCard.fusion [hunk 8]
```diff
@@ -1,87 +1,42 @@
...
+        <meta @key="description" name="twitter:description" content={props.description} @if.hasDescription={props.description}/>
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/TwitterCard.fusion]: `@if.has`

File: Resources/Private/Fusion/Metadata/TwitterCard.fusion [hunk 9]
```diff
@@ -1,87 +1,42 @@
...
+        <meta @key="creator" name="twitter:creator" content={props.creator} @if.hasCreator={props.creator}/>
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/TwitterCard.fusion]: `@if.has`

File: Resources/Private/Fusion/Metadata/TwitterCard.fusion [hunk 10]
```diff
@@ -1,87 +1,42 @@
...
+        <meta @key="url" name="twitter:url" content={props.url}/>
+        <meta @key="image" name="twitter:image" content={props.image} @if.hasImage={props.image}/>
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/TwitterCard.fusion]: `@if.has`

File: Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion [hunk 11]
```diff
@@ -1,90 +1,36 @@
...
+
+    renderer = afx`
+        <meta @key="type" property="og:type" content={props.type}/>
+        <meta @key="title" property="og:title" content={props.title}/>
+        <meta @key="siteName" property="og:site_name" content={props.title}/>
+        <meta @key="locale" property="og:locale" content={props.locale} @if.set={props.locale}/>
+        <meta @key="description" property="og:description" content={props.description} @if.set={props.description}/>
+        <meta @key="url" property="og:url" content={props.url}/>
+        <Neos.Fusion:Join @key="image" @if.set={props.image}>
+            <meta @key="image" property="og:image" @children="attributes.content">
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion]:
```suggestion
            <meta @key="image" property="og:image">
```

File: Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion [hunk 12]
```diff
@@ -1,90 +1,36 @@
...
+                <Neos.Neos:ImageUri asset={props.image} preset='Neos.Seo:OpenGraph'/>
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion]:
```suggestion
                <Neos.Neos:ImageUri asset={props.image} preset='Neos.Seo:OpenGraph'  @path='attributes.content'/>
```

File: Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion [hunk 13]
```diff
@@ -1,90 +1,36 @@
...
+            </meta>
+            <meta @key="width" property="og:image:width" content={props.image.width}/>
+            <meta @key="height" property="og:image:height" content={props.image.height}/>
+            <meta @key="alt" property="og:image:alt" content={props.image.label}/>
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/OpenGraphMetaTags.fusion]:
```suggestion
            <meta @key="alt" property="og:image:alt" content={props.image.label} @if.set={props.image.label}/>
```

File: Resources/Private/Fusion/Metadata/TwitterCard.fusion [hunk 14]
```diff
@@ -1,87 +1,42 @@
...
-        @if.isSet = ${Configuration.setting('Neos.Seo.twitterCard.siteHandle') != null}
+prototype(Neos.Seo:TwitterCard) < prototype(Neos.Fusion:Component) {
+    type = ${q(node).property('twitterCardType') || 'summary'}
+    title = ${q(node).property('twitterCardTitle') || q(node).property('titleOverride') || q(node).property('title')}
+    description = ${q(node).property('twitterCardDescription') || q(node).property('metaDescription')}
+    siteHandle = ${Configuration.setting('Neos.Seo.twitterCard.siteHandle')}
+    creator = ${q(node).property('twitterCardCreator')}
+    url = Neos.Neos:NodeUri {
+        node = ${node}
+        absolute = TRUE
```
Comment (jonnitto) [inline Resources/Private/Fusion/Metadata/TwitterCard.fusion]:
```suggestion
        absolute = true
```

Comment (jonnitto) [reply Resources/Private/Fusion/Metadata/TwitterCard.fusion]: I think this is the last one ☝️ 😄

**Pull request merged**